Project Proposal: Audit Logging SIG (#2409)
open-telemetry/community: projects/audit-logging.md

Audit logging describes the capability of capturing audit-trail relevant events of a system to meet compliance requirements. Such events may originate from the infrastructure (e.g. a Kubernetes cluster) up to the application-level. It is a capability that is particularly relevant for providers of enterprise software.

Unlike regular application logs, audit logs are usually subject to long retention periods and software providers must guarantee their completeness (i.e. guarantee of delivery).

Examples of audit logs include: (see [Appendix B: Sample Audit Log Events])
- failed login attempts
- permission changes (e.g. of a service account or application user)
- accessing sensitive information
- modification of data

### Current challenges

OpenTelemetry does not have a good solution for audit logging

- no semantic conventions for audit logs in OTel
- OTel APIs/SDKs do not provide feedback to the application level whether data (in particular logs) have been successfully delivered to a remote endpoint. To guarantee delivery, either the SDK has to give those guarantees, or provide feedback to the application so that it can take care of guaranteed delivery itself.

> - OTel collectors may lose audit logs in transit (i.e. no guarantee of delivery)

Reviewer: Does "OTel collectors" mean the "OpenTelemetry Collector" (https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification#project-naming) or "any collectors that can handle OpenTelemetry data (whether OTLP or something else)"?

Author: @reyang This is indeed supposed to refer to the OpenTelemetry Collector. I've rephrased this particular sentence to  "OTel Collector instances may lose audit logs in transit" and adjusted other occurrences of _OTel collectors_ as well. https://github.com/open-telemetry/community/pull/2409/commits/066501b92566ce770eb79f5633d2c1e5166aa9a5

### Required staffing

* Project lead: SAP (name tbd)
* Sponsors:
  - @reyang
  - tbd
* GC liaison: @svrnm
* Engineers:
  * SAP will provide a prototype in two languages (tbd; likely two of Java, JavaScript, Go)
* Maintainers/approvers: tbd

## Timeline

TBD based on community involvement.

## Labels

- audit-logging (tbc)

## Project Board

TODO: add link

## SIG Meetings and Other Info

TODO: add information

## Appendix

### Appendix A: Guarantee of Delivery

In the context of this document, guarantee of delivery describes the ability of delivering audit logs from source to destination through OTel means while ensuring that all such signals arrive at the destination and/or providing the source with a means to handle failed delivery.

Messaging protocols that support different levels of delivery guarantees may refer to this behavior as _at least once_ or _exactly once_, as opposed to _at most once_.

We assume that every component that is involved in the delivery of audit logs from source to destination must support guarantee of delivery individually, rather than assuming that this ability can be provided by e.g. only the collector or SDK.

> The implications of guarantee of delivery can be illustrated with an example consisting of a workload, an OTel collector, and a durable storage. The workload acts as the source and produces audit logs via the OTel SDK/API. It writes the data via OTLP to the OTel collector. The OTel collector is configured to export audit logs to a durable storage that acts as the destination such as an S3 bucket.

Reviewer: Minor, it is a bit weird to see "SDK/API", I think folks normally use "API/SDK".

The following implications would apply:

- workload produces an audit-relevant event:

  The workload emits the event via the OTel SDK/API. It may wait for acknowledgement of receipt from the collector before proceeding. If the event is rejected or receipt is not acknowledged in time, the workload or SDK may act accordingly, e.g. retry, rollback a database transaction, inform the user, etc.

- OTel collector receives the event:

>   To ensure that the event is not lost even if the collector process is terminated or crashes, the collector may need to persist the event before acknowledging receipt to the workload or SDK. If the event cannot be persisted, receipt must be rejected.

Reviewer: What is the expectation if the collector instance disappeared (e.g., the machine running the collector exploded / was stolen)?

Reviewer 2: I think this is the most tricky part, or to put it in a question: do we need guarantee of delivery between 2 components (workload->collector,collector->S3) or end-to-end (workload->S3)?

I would assume "end-to-end" except the collector can guarantee that data is persisted according the auditing requirements

Author: > (e.g., the machine running the collector exploded / was stolen)
_If_ the solution for audit logging with OTel meant that the OTel Collector had an own persistence, I would argue that theft/explosion/etc. are rather in the responsibility of Operations in terms of configuring said persistence such that it is resilient "enough".
Or to make this more concrete: _if_ for example something such as the [storage extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage) was used, Operations would need to make sure that the database/file/redis storage runs in an HA mode.
I'm stressing the _if_ here, because I think is a detail that the SIG should work out. Or do you think that's something that should rather be clarified upfront?

Reviewer: > I'm stressing the _if_ here, because I think is a detail that the SIG should work out. Or do you think that's something that should rather be clarified upfront?

I suggest that we leave this for the SIG to figure out. In the OTEP, I suggest that we avoid "guaranteed delivery" and use something like "certain degree/level of data delivery guarantee". Not a blocker for this PR though (I'm good with the current version).

Reviewer 2: +1 for what @reyang wrote. I think it is good to have this in the appendix and some wording around it, since there is many people (including myself) who have a superficial knowledge around audit logs, so it helps to contextualize and understand what this is all about

- OTel collector exports the event:

  Once the event is exported and the target (i.e. S3) acknowledges receipt, the event can dropped from the collector's persistence.

> - S3 receives the event:

Reviewer: S3 -> the target (i.e. S3)

Author: good point, makes it more consistent. https://github.com/open-telemetry/community/pull/2409/commits/066501b92566ce770eb79f5633d2c1e5166aa9a5

  Acknowledges receipt after persisting the event.

  Note that this is outside the scope of the OTel. More general, when using OTel for audit logging purposes, it's the users (e.g. Ops) responsibility to configure a suitable export target.

Note that this example may contain implementation details for illustration purposes. The actual implementation may differ as long as the requirements are met.

The example is kept simple for illustration purposes. Many edge cases need to be discussed by the SIG, such as batch-sending of signals or handling of multiple export targets.

It may turn out that all OTel receivers, processors, or exporters can be made compatible with guarantee of delivery for audit logging purposes.

> ### Appendix B: Sample Audit Log Events

Author: prime example of how words can lead to completely different expectations depending on one's background 😄  I've changed it to "Examples of Audit Log Events". https://github.com/open-telemetry/community/pull/2409/commits/066501b92566ce770eb79f5633d2c1e5166aa9a5

# Audit Logging

## Background and description

Audit logging describes the capability of capturing audit-trail relevant events of a system to meet compliance requirements. Such events may originate from the infrastructure (e.g. a Kubernetes cluster) up to the application-level. It is a capability that is particularly relevant for providers of enterprise software.

> Unlike regular application logs, audit logs are usually subject to long retention periods and software providers must guarantee their completeness (i.e. guarantee of delivery).

Reviewer: Good points! In addition, these are something we might want to consider:
1. Audit logs might be considered as a critical part of the business, which could result in a different API design strategy - for example, audit logging might require a different API behavior, if the information provided by the caller is invalid, the API might throw exception instead of failing silently and move on.
2. Audit log might require some sensitive information without redaction due to the regulation requirements (e.g. user identity and client IP address).
3. The data path could require higher level of access control or privilege.

Author: @reyang thanks for mentioning these points.

Especially the API behavior is something that we had thought about initially. However, when we first pitched audit logging [on Slack](https://cloud-native.slack.com/archives/CJFCJHG4Q/p1719840258022029), we received the following comment from Ted Young:

> As a rule, the OpenTelemetry API never throws an exception. I understand why you might want this, though it is not present in many audit logging systems, which use regular loggers. So a strong case would have to be made on this particular point.

Based on this initial feedback, we decided to file this SIG proposal without proposing API changes.

Examples of audit logs include:
- permission changes (e.g. of a service account or application user)
- modification of data
- accessing sensitive information
- failed login attempts

### Current challenges

> Audit Logging is currently not within the scope of OpenTelemetry

Reviewer: I have a different perspective here, I think OpenTelemetry doesn't have a good solution for audit logging, but audit logging is definitely within the current scope (it is still a type of logs / timestamped events).

Reviewer: 👍

> - no semantic conventions for audit logs in OTEL

Reviewer 3: ```suggestion
- There aren't currently any semantic conventions designed specifically for audit logs in OTEL
```

> - no semantic conventions for audit logs in OTEL

Author: ```suggestion
- no semantic conventions for audit logs in OTel
```

> - OTEL APIs/SDKs do not provide feedback to the application level whether data (in particular logs) have been successfully delivered to a remote endpoint. To guarantee delivery, either the SDK has to give those guarantees, or provide feedback to the application so that it can take care of guaranteed delivery itself.

Reviewer: https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification#project-naming

```suggestion
- OTel APIs/SDKs do not provide feedback to the application level whether data (in particular logs) have been successfully delivered to a remote endpoint. To guarantee delivery, either the SDK has to give those guarantees, or provide feedback to the application so that it can take care of guaranteed delivery itself.
```

> - OTEL collectors may lose audit logs in transit (i.e. no guarantee of delivery)

Author: ```suggestion
- OTel collectors may lose audit logs in transit (i.e. no guarantee of delivery)
```

### Goals, objectives, and requirements

> The goal of this project is to make OTEL fit for audit logging purposes that meet compliance requirements of enterprise software providers, in particular:

Author: ```suggestion
The goal of this project is to make OTel fit for audit logging purposes that meet compliance requirements of enterprise software providers, in particular:
```

- REQ-CONV-01: Semantic conventions for application-level audit logs are defined
- REQ-CONV-02: Semantic conventions for infrastructure-level audit logs are defined

> - REQ-APPL-01: Guaranteed delivery of audit logs exported via OpenTelemetry SDK.

Reviewer: nit: remove the "." or add "." to the other items

> - REQ-PIPE-01: OTEL collector must provide guaranteed delivery of audit logs, including when its process is interrupted

Reviewer: I think we should clarify/define "guaranteed delivery".

Author: @reyang I've added a chapter to the appendix on guarantee of delivery in the context of audit logging: https://github.com/open-telemetry/community/pull/2409/files#diff-736e6b0ae9ae655b78d9ba007d08592071abb6cc1ef64d7893ff81642c8ec734R79-R113.

## Deliverables

- semantic convention for audit logs

> - extend OTEL APIs/SDKs for audit logging purposes (in collaboration with the respective SIG)
> - extend OTEL collector for audit logging purposes (in collaboration with the respective SIG)

Author: ```suggestion
- extend OTel APIs/SDKs for audit logging purposes (in collaboration with the respective SIG)
- extend OTel collector for audit logging purposes (in collaboration with the respective SIG)
```

## Staffing / Help Wanted

The following vendors are interested in improving this area:

> - SAP

Reviewer: You can put me here, I'm from Microsoft, I'm currently a Technical Committee member of OpenTelemetry, and I work on audit logs as part of my job (I don't own the Microsoft audit logging story, but I work with the team closely).

Other vendors are invited to join the discussion.

### Required staffing

* Project lead: SAP (name tbd)

> * Sponsors: tbd

Reviewer: I can sponsor this.

* GC liaison: tbd
* Engineers:

>   * SAP will provide a prototype in two languages (tbd; likely two of Java, JavaScript, Go)

Reviewer: I think we need prototype in two parts:
1. API/SDK - this is where we need three programming languages IIRC.
2. OTel Collector - higher guarantee on data delivery (completeness, integrity, latency, etc.), data path security.

Author: Thanks for pointing this out! It's clear to us, but I'll make this clearer in the doc...

# Audit Logging

## Background and description

Audit logging describes the capability of capturing audit-trail relevant events of a system to meet compliance requirements. Such events may originate from the infrastructure (e.g. a Kubernetes cluster) up to the application-level. It is a capability that is particularly relevant for providers of enterprise software.

Unlike regular application logs, audit logs are usually subject to long retention periods and software providers must guarantee their completeness (i.e. guarantee of delivery).

Examples of audit logs include:
- permission changes (e.g. of a service account or application user)
- modification of data
- accessing sensitive information
- failed login attempts

### Current challenges

Audit Logging is currently not within the scope of OpenTelemetry

> - no semantic conventions for audit logs in OTel

Author: @svrnm our experience has shown that in order to analyze audit logs at scale, it is important to define an (extensible) event catalog. The event catalog standardizes audit log events across workloads/produces. For example, our internal event catalog currently consists of 50+ such events. Ideally, such a catalog would be part of semantic conventions.

To make this more tangible, I've added some examples to the appendix of the document:
https://github.com/open-telemetry/community/pull/2409/files#diff-736e6b0ae9ae655b78d9ba007d08592071abb6cc1ef64d7893ff81642c8ec734R115-R192

> - no semantic conventions for audit logs in OTel

Reviewer 4: another examples from the security world is https://github.com/ocsf.

> - no semantic conventions for audit logs in OTel

Reviewer 2: thanks @mlenkeit. Makes it much clearer

The `metadata` looks like attributes that would be covered by other semantic conventions (e.g. there is a [`log.record.id`](https://opentelemetry.io/docs/specs/semconv/attributes-registry/log/#log-record-attributes) for the `metadata.id`, the timestamp of course and some of the other ones (e.g. for `k8sCluster` we have [`k8s.cluster.name`](https://opentelemetry.io/docs/specs/semconv/attributes-registry/k8s/#kubernetes-attributes). So I would assume here it is more about re-using and extending certain other domains that are not unique to "audit logs"

For the `event` and `data` examples you gave, I would argue that they are not "semantic conventions for audit logs" but "semantic conventions for log types that typically require the strict requirements of auditing". What do I mean by that: if we talk about "semantic conventions for audit logs" I think about a namespace called `audit.` that holds attributes that are specific to the business logic of audit logging, like a signature that helps to tamperproof the log line, or maybe even meta information under which regulation this log is required to be an "audit log"
In contrast "semantic conventions for log types that typically require the strict requirements of auditing" are their own namespaces like the "UserLoginFailure" example would fall into a "authentication" or "auth" namespace, with "auth.login.method" or "auth.login.failureReason" as potential attributes, `event.name` being set to `auth.login.failure` or something.

I am just making those things up to exemplify the difference, they will probably take a different form or shape eventually, so to make a long story short, here is a suggestion to rephrase:

```suggestion
- no semantic conventions for audit logs in OTel
- no semantic conventions for log types that typically rquire the strict requirements of auditing, like authentication, authorization and data changes
```

@renewelches thanks for calling out OCSF, if I remember correctly there were conversations in the past between OTel and OCSF, cc @lmolkova

> - no semantic conventions for audit logs in OTel

Author: Regarding `metadata`, I fully agree: Most of these attributes are already covered by semconv. We may identify additional attributes in SIG meetings though, depending on the experience/requirement of other contributors/companies.

I understand how "semantic conventions for audit logs" can be misleading. To me, the suggestion that you made has a notion of particularly describe logs that are "already there" (e.g. events emitted by a K8s cluster) and can be considered relevant for audit purposes. Especially in enterprise software, it's common that applications produce logs that are specifically mean to be audit logs (and nothing else). To me, it' s important that we find wording that covers these two types that we do have.

How about the following?

```suggestion
- no semantic conventions for representing and identifying audit trail-relevant events in OTel (like authentication, authorization or modification of

> - no semantic conventions for audit logs in OTel

Reviewer 5: +1 to looking into OCSF for security events and borrowing relevant semantic conventions from there.

> - no semantic conventions for audit logs in OTel

Reviewer 6: @mlenkeit security semconv SIG is also working on introducing security related attributes to the semconv, we have currently discussion around [user authentication](https://github.com/open-telemetry/semantic-conventions/pull/1456). I feel that this SIG and security SIG might have some common tasks/discussions

- OTel APIs/SDKs do not provide feedback to the application level whether data (in particular logs) have been successfully delivered to a remote endpoint. To guarantee delivery, either the SDK has to give those guarantees, or provide feedback to the application so that it can take care of guaranteed delivery itself.

> - OTel collectors may lose audit logs in transit (i.e. no guarantee of delivery)

Author: @svrnm I've added a chapter to the appendix on guarantee of delivery in the context of audit logging: https://github.com/open-telemetry/community/pull/2409/files#diff-736e6b0ae9ae655b78d9ba007d08592071abb6cc1ef64d7893ff81642c8ec734R79-R113.

I think this capability should be built such that it can be enabled for audit logs if required while other signals continue to be processed/delivered just like they are today.

### Goals, objectives, and requirements

The goal of this project is to make OTel fit for audit logging purposes that meet compliance requirements of enterprise software providers, in particular:

- REQ-CONV-01: Semantic conventions for application-level audit logs are defined
- REQ-CONV-02: Semantic conventions for infrastructure-level audit logs are defined
- REQ-APPL-01: Guaranteed delivery of audit logs exported via OpenTelemetry SDK
- REQ-PIPE-01: OTel collector must provide guaranteed delivery of audit logs, including when its process is interrupted

## Deliverables

- semantic convention for audit logs
- extend OTel APIs/SDKs for audit logging purposes (in collaboration with the respective SIG)
- extend OTel collector for audit logging purposes (in collaboration with the respective SIG)

## Staffing / Help Wanted

The following vendors are interested in improving this area:
- SAP (@mlenkeit, @FWinkler79)
- Microsoft (@reyang)

Other vendors are invited to join the discussion.

### Required staffing

* Project lead: SAP (name tbd)
* Sponsors:
  - Reiley Yang (@reyang)
  - tbd

> * GC liaison: tbd

Reviewer 2: I volunteer for being the liaison if this is going to become a SIG

Author: Thank you! I've updated the doc accordingly https://github.com/open-telemetry/community/pull/2409/commits/3876a315eaad74539bc52110f8a430a985acdb55.
